### PR TITLE
scripts/g.extension.all: fix reinstall multi-addons

### PR DESCRIPTION
**Pull request by tmszi**
**Describe the bug**
Reinstall multi-addon e.g. wx.metadata (which contains multiple individual modules), doesn't work.

**To Reproduce**

1. Install wx.metadata addon `g.extension wx.metadata`
2. Reinstall all installed addons `g.extension.all -f`
3. See error

```
------------------------------------------------------------
Reinstalling extension <v.info.iso>...
------------------------------------------------------------
WARNING: Extension <v.info.iso> already installed. Re-installing...
Fetching <v.info.iso> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/vector/v.info.iso' doesn't exist
ERROR: GRASS Addons <v.info.iso> not found
ERROR: Unable to process extension:v.info.iso
------------------------------------------------------------
Reinstalling extension <t.info.iso>...
------------------------------------------------------------
WARNING: Extension <t.info.iso> already installed. Re-installing...
Fetching <t.info.iso> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/temporal/t.info.iso' doesn't exist
ERROR: GRASS Addons <t.info.iso> not found
ERROR: Unable to process extension:t.info.iso
------------------------------------------------------------
Reinstalling extension <r.info.iso>...
------------------------------------------------------------
WARNING: Extension <r.info.iso> already installed. Re-installing...
Fetching <r.info.iso> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/raster/r.info.iso' doesn't exist
ERROR: GRASS Addons <r.info.iso> not found
ERROR: Unable to process extension:r.info.iso
------------------------------------------------------------
Reinstalling extension <m.csw.update>...
------------------------------------------------------------
WARNING: Extension <m.csw.update> already installed. Re-installing...
Fetching <m.csw.update> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/misc/m.csw.update' doesn't exist
ERROR: GRASS Addons <m.csw.update> not found
ERROR: Unable to process extension:m.csw.update
------------------------------------------------------------
Reinstalling extension <g.gui.metadata>...
------------------------------------------------------------
WARNING: Extension <g.gui.metadata> already installed. Re-installing...
Fetching <g.gui.metadata> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/general/g.gui.metadata' doesn't exist
ERROR: GRASS Addons <g.gui.metadata> not found
ERROR: Unable to process extension:g.gui.metadata
------------------------------------------------------------
Reinstalling extension <g.gui.cswbrowser>...
------------------------------------------------------------
WARNING: Extension <g.gui.cswbrowser> already installed. Re-installing...
Fetching <g.gui.cswbrowser> from GRASS GIS Addons repository (be
patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/general/g.gui.cswbrowser' doesn't exist
ERROR: GRASS Addons <g.gui.cswbrowser> not found
ERROR: Unable to process extension:g.gui.cswbrowser
------------------------------------------------------------
Reinstalling extension <db.csw.run>...
------------------------------------------------------------
WARNING: Extension <db.csw.run> already installed. Re-installing...
Fetching <db.csw.run> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/db/db.csw.run' doesn't exist
ERROR: GRASS Addons <db.csw.run> not found
ERROR: Unable to process extension:db.csw.run
------------------------------------------------------------
Reinstalling extension <db.csw.harvest>...
------------------------------------------------------------
WARNING: Extension <db.csw.harvest> already installed. Re-installing...
Fetching <db.csw.harvest> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/db/db.csw.harvest' doesn't exist
ERROR: GRASS Addons <db.csw.harvest> not found
ERROR: Unable to process extension:db.csw.harvest
------------------------------------------------------------
Reinstalling extension <db.csw.admin>...
------------------------------------------------------------
WARNING: Extension <db.csw.admin> already installed. Re-installing...
Fetching <db.csw.admin> from GRASS GIS Addons repository (be patient)...
svn: E170000: URL 'https://github.com/OSGeo/grass-addons/branches/grass8/src/db/db.csw.admin' doesn't exist
ERROR: GRASS Addons <db.csw.admin> not found
ERROR: Unable to process extension:db.csw.admin
```

**Expected behavior**
Reinstall multi-addon e.g. wx.metadata (which contains multiple individual modules), should be work.

**System description:**

- GRASS GIS version main, realeasebranch_8_0, releasebranch_7_8

**Additional context**
Affected is reinstallation of the multi-addons (which contains multiple individual modules) e.g. wx.metadata, r.connectivity etc.

### CODE REVIEW

File: scripts/g.extension.all/g.extension.all.py [hunk 1]
```diff
@@ -77,6 +88,124 @@ def get_extensions():
...
+                ),
+            )
+        return response
+
+    except HTTPError as err:
+        if err.code == 404:
+            gscript.fatal(
+                _(
+                    "The download of the modules.xml file "
+                    "from the osgeo server wasn't successful. "
```
Comment (neteler) [inline scripts/g.extension.all/g.extension.all.py:143]:
```suggestion
                    "from the server grass.osgeo.org was not successful. "
```

Comment (neteler) [reply scripts/g.extension.all/g.extension.all.py:143]: I find "osgeo" somehow confusing in this context, and the URL comes anyway in the next line...

File: scripts/g.extension.all/g.extension.all.py [hunk 2]
```diff
@@ -77,6 +88,124 @@ def get_extensions():
...
+        )
+
+
+def find_addon_name(addons):
+    """Find correct addon name if addon is complex type addon
+    e.g. wx.metadata has multiple modules g.gui.cswbrowser etc.
+
+    for g.gui.cswbrowser module is return wx.metadata addon name
+    for r.info.iso module is return wx.metadata addon name
+    etc.
```
Comment (neteler) [inline scripts/g.extension.all/g.extension.all.py:170]:
```suggestion
    """Find correct addon name if addon is a multi-addon,
    e.g. wx.metadata contains multiple modules g.gui.cswbrowser etc.

    Examples:
    - for the g.gui.cswbrowser module the wx.metadata addon name is returned
    - for the i.sentinel.download module the i.sentinel.download addon name is returned 
    etc.
```

File: scripts/g.extension.all/g.extension.all.py [hunk 3]
```diff
@@ -77,6 +88,124 @@ def get_extensions():
...
+
+    :param list addons: list of individual addons modules to reinstall
+
+    :return set result: set of unique addons names to reinstall
```
Comment (neteler) [inline scripts/g.extension.all/g.extension.all.py:174]:
```suggestion

    :param list addons: list of individual addon modules to be reinstalled

    :return set result: set of unique addon names to be reinstalled
```

**Comment by tmszi** [comment]
Rebase https://github.com/OSGeo/grass/commit/2b383979a38a676851c082c2b13f777d6a5f366a.

File: scripts/g.extension.all/g.extension.all.py [hunk 4]
```diff
@@ -77,6 +87,127 @@ def get_extensions():
...
+                        url=url,
+                        code=response.code,
+                        desc=desc,
+                    ),
+                ),
+            )
+        if response_format not in response.getheader("Content-Type"):
+            gscript.fatal(
+                _(
+                    "Wrong downloaded file format. "
```
Comment (veroandreo) [inline scripts/g.extension.all/g.extension.all.py:127]:
```suggestion
                    "Wrong file format downloaded. "
```

File: scripts/g.extension.all/g.extension.all.py [hunk 5]
```diff
@@ -77,6 +87,127 @@ def get_extensions():
...
+
+
+def find_addon_name(addons):
+    """Find correct addon name if addon is a multi-addon
+    e.g. wx.metadata contains multiple modules g.gui.cswbrowser etc.
+
+    Examples:
+    - for the g.gui.cswbrowser module the wx.metadata addon name is
+    returned
+    - for the i.sentinel.download module the i.sentinel.download addon
```
Comment (veroandreo) [inline scripts/g.extension.all/g.extension.all.py:170]:
```suggestion
    - for the i.sentinel.download module the i.sentinel addon
```
The parent multi-addon is to be called, right?

**Review by veroandreo — COMMENTED** [review]
some minor suggestions in messages

File: scripts/g.extension.all/g.extension.all.py (hunk 5 above)
Comment (tmszi) [reply scripts/g.extension.all/g.extension.all.py:170]: Yes, you are right.

File: scripts/g.extension.all/g.extension.all.py (hunk 4 above)
Comment (tmszi) [reply scripts/g.extension.all/g.extension.all.py:127]: Thanks for review.

**Comment by tmszi** [comment]
> There is still an (unrelated) glitch in the OSGeo4W CI:
> 
> ```
> Run .github/workflows/test_simple.bat 'C:\OSGeo4W\opt\grass\grass80.bat'
> 'C:\OSGeo4W\opt\grass\grass80.bat' is not recognized as an internal or external command,
> 
> D:\a\grass\grass>set grass=C:\OSGeo4W\opt\grass\grass80.bat 
> operable program or batch file.
> 
> 'C:\OSGeo4W\opt\grass\grass80.bat' is not recognized as an internal or external command,
> D:\a\grass\grass>rem Test execution of binary command 
> operable program or batch file.
> 
> 'C:\OSGeo4W\opt\grass\grass80.bat' is not recognized as an internal or external command,
> D:\a\grass\grass>call C:\OSGeo4W\opt\grass\grass80.bat --tmp-location EPSG:4326 --exec g.region res=0.1 -p 
> operable program or batch file.
> ```
> 
> I wonder (since it worked elsewhere; after merge of #2074 yesterday) if this PR simply needs a rebase?

This commit https://github.com/OSGeo/grass/commit/029cd8ea8e9c98b575f57f148b1552c8821a0494 has led to a change https://github.com/tmszi/grass/runs/4792759604?check_suite_focus=true#step:6:36500 C:\OSGeo4W\opt\grass\grass80.bat -> C:\OSGeo4W\opt\grass\grass81.bat.

**Comment by neteler** [comment]
I have tested it with the freshly compiled G81, `g.extension.all -f` runs (on Linux).

An (unrelated?) slighlty confusing message occurs with the multi-addons:

```
GRASS nc_spm_08_grass7/user1:grass_main > g.extension.all -f

...
Fetching <i.sentinel> from GRASS GIS Addons repository (be patient)...
Compiling...
Installing...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'i.sentinel': Unable to fetch
         interface description for command '<i.sentinel>'.

         Details: <[Errno 2] No such file or directory: 'i.sentinel'>
Installation of <i.sentinel> successfully finished
...
```

But perhaps that's for a separate PR.

**Comment by tmszi** [comment]
> I have tested it with the freshly compiled G81, `g.extension.all -f` runs (on Linux).
> 
> An (unrelated?) slighlty confusing message occurs with the multi-addons:
> 
> ```
> GRASS nc_spm_08_grass7/user1:grass_main > g.extension.all -f
> 
> ...
> Fetching <i.sentinel> from GRASS GIS Addons repository (be patient)...
> Compiling...
> Installing...
> Updating extensions metadata file...
> Updating extension modules metadata file...
> WARNING: No metadata available for module 'i.sentinel': Unable to fetch
>          interface description for command '<i.sentinel>'.
> 
>          Details: <[Errno 2] No such file or directory: 'i.sentinel'>
> Installation of <i.sentinel> successfully finished
> ...
> ```
> 
> But perhaps that's for a separate PR.

It is just a warning message not bug in this case (I will try to explain below, which is justified there) and it is not related to this PR.

Multi-addons e.g. i.sentinel contains [multiple individual addons](https://github.com/OSGeo/grass-addons/tree/grass8/src/imagery/i.sentinel). Every individual addons i.sentinel.coverage, i.sentinel.download, 
i.sentinel.import, i.sentinel.mask, i.sentinel.parallel.download, i.sentinel.preproc has binary Python file and metadata is returned via this [function in g.extension module](https://github.com/OSGeo/grass/blob/main/scripts/g.extension/g.extension.py#L1294).

But **root addon directory i.sentinel** doesn't have Python binary, contain only root addon [manual html page](https://github.com/OSGeo/grass-addons/blob/grass8/src/imagery/i.sentinel/i.sentinel.html), which is [installed](https://github.com/OSGeo/grass-addons/blob/grass8/src/imagery/i.sentinel/Makefile#L17) and therefore users will see this warning message.

**Review by neteler — APPROVED** [review]

**Pull request merged**